CONTRIBUTING.md

```diff
@@ -209,6 +209,72 @@ Before submitting your changes, run the test suite to make sure that nothing is
+        be internal rather than general purpose. We have no plans to
+        [semver](https://semver.org/) their APIs or accept feature requests
+        for them. We publish them to maven central because they are
+        dependencies of our plugin test framework, high level rest client, and
+        jdbc driver but.
+* modules:  Features that are shipped with Elasticsearch by default but are not
+        built in to the server. We typically separate features from the server
+        because they require permissions that we don't believe *all* of
+        Elasticsearch should have or because they depend on libraries that we
+        don't believe *all* of Elasticsearch should depend on. For example,
+        reindex requires the `connect` permission so it can perform
+        reindex-from-remote but we don't believe that the *all* of
+        Elasticsearch should have the "connect". For another example, Painless
+        is implemented using antlr4 and asm and we don't believe that *all*
+        of Elasticsearch should have access to them.
+* plugins:  Officially supported plugins to Elasticsearch. We decide that a
+        feature should be a plugin rather than built in to the Elasticsearch
+        server or shipped as a module because we feel that it is only important
+        to a subset of users, especially if it requires extra dependencies. The
+        canonical example of this is the ICU analysis plugin. It is important
+        for folks who want the fairly language neutral ICU analyzer but the
+        library to implement the analyzer is 11MB.
+* qa:  Honestly this is kind of in flux and we're not 100% sure where we'll end
+        up. Right now the directory contains
+    * Tests that require multiple modules or plugins to work
+    * Tests that form a cluster made up of multiple versions of Elasticsearch
+            like full cluster restart, rolling restarts, and mixed version
+            tests
+    * Tests that test the Elasticsearch clients in "interesting" places like
+            the `wildfly` project.
+    * Tests that test Elasticsearch in funny configurations like with ingest
+            disabled
+    * Tests that need to do strange things like install plugins that thrown
+            uncaught `Throwable`s or add a shutdown hook
+        But we're not convinced that all of these things *belong* in the qa
+        directory. We're fairly sure that "tests that require multiple modules
+        or plugins to work" should just pick a "home" plugin. We're fairly
+        sure that the multi-version tests *do* belong in qa. Beyond that, we're
+        not sure. If you want to add a new thing qa module, ask around and
+        you'll get several opinions.
+* server:  The server component of Elasticsearch that contains all of the
+        modules and plugins. Right now things like the high level rest client
+        depend on the server but we'd like to fix that in the future.
```

Comment:
Another suggestion of how you might avoid "we": "there are plans to fix..." would work well I think.